asteroid_filterbanks/pcen.py

```diff
@@ -0,0 +1,270 @@
+    """
+    Computes the exponential moving average of an sequential input.
+
+    Influenced by leaf-audio's tensorflow implementation
+        https://github.com/google-research/leaf-audio/blob/7ead2f9fe65da14c693c566fe8259ccaaf14129d/leaf_audio/postprocessing.py#L27
+
+    See license here
+        https://github.com/google-research/leaf-audio/blob/master/LICENSE
+    """
+
+    def __init__(
+        self,
+        smooth: float = 0.04,
+        per_channel: bool = False,
+        n_channels: int = 1,
+        trainable: bool = False,
+    ):
+        super().__init__()
+        if per_channel:
+            weights = torch.full((n_channels,), fill_value=smooth)
+        else:
+            weights = torch.full((1,), fill_value=smooth)
+
+        self.set_weights(data=weights, trainable=trainable)
+
+    def set_weights(self, data: torch.Tensor, trainable: bool):
+        self.weights = nn.Parameter(data, requires_grad=trainable)
+        self.trainable = trainable
+
+    def forward(
+        self, mag_spec: torch.Tensor, initial_state: Optional[torch.Tensor]
+    ) -> Tuple[torch.tensor, torch.Tensor]:
+        if initial_state is None:
+            initial_state = mag_spec[:, :, :, 0].unsqueeze(-1)
+
+        weights = torch.clamp(self.weights, 0.0, 1.0)
+        accumulator = initial_state.transpose(1, -1)
+        out = None
+        for x in torch.split(mag_spec.transpose(1, -1), 1, dim=1):
+            accumulator = weights * x + (1.0 - weights) * accumulator
+            accumulator = accumulator
+            if out is None:
+                out = accumulator.transpose(1, -1)
+            else:
+                out = torch.cat((out, accumulator.transpose(1, -1)), dim=-1)
+        return out, accumulator.transpose(1, -1)
+
+
+class TrainableParameters(TypedDict):
+    alpha: bool
+    delta: bool
+    root: bool
+    smooth: bool
+
+
+class _PCEN(nn.Module):
+    def __init__(
+        self,
+        alpha: float = 0.96,
+        delta: float = 2.0,
+        root: float = 2.0,
+        floor: float = 1e-6,
+        smooth: float = 0.04,
+        n_channels: int = 1,
+        trainable: Union[bool, TrainableParameters] = False,
+        per_channel_smoothing: bool = False,
+    ):
+        super().__init__()
+
+        if trainable is True or trainable is False:
+            trainable = TrainableParameters(
+                alpha=trainable, delta=trainable, root=trainable, smooth=trainable
+            )
+
+        self.trainable = trainable
+        self.n_channels = n_channels
+
+        self.alpha = nn.Parameter(
+            torch.full((self.n_channels,), fill_value=alpha), requires_grad=self.trainable["alpha"]
+        )
+        self.delta = nn.Parameter(
+            torch.full((self.n_channels,), fill_value=delta), requires_grad=self.trainable["delta"]
+        )
+        self.root = nn.Parameter(
+            torch.full((self.n_channels,), fill_value=root), requires_grad=self.trainable["root"]
+        )
```

Comment:
Use `torch.ones(...) * x` and refold in single lines.

Reply:
![image](https://user-images.githubusercontent.com/1007064/113984866-d8ab3480-984b-11eb-9d13-a32ae69c296c.png)

From experimentation, `full` is 3x as fast.

Reply:
Also, the formatting is automatically applied by black.